### ISSUE TITLE
Receiving does not update cost, tax and discount?

### ISSUE THREAD

**Issue by bkrajendra** [jekkos/opensourcepos#544]
What I believe with receiving is that its supposed to be is "Purchase". 
And default behaviour of Purchase is to update stock.
In addition to this its necessary to update tax rate, Cost, Selling price and discount related to that item.

But this is not happening here. Is it done intentionally or am i missing something.
When I put new Cost and quantity in receiving form, it only updates stock. New cost is visible in receipt but this new cost is not updated in inventory.

**Comment by daN4cat** [comment]
Receiving is done after a purchasing action.
Receiving is an action on stock.

If you feel the feature doesn't fit the standard receiving action happy for you to contribute improvements. We need help.

**Comment by RamkrishnaMondal** [comment]
I also believe that, receiving module should be replaced with a detailed "purchases" module, but we need help from other developers. We can't expect jekkos and daN4cat to do everything for us.

**Comment by daN4cat** [comment]
Indeed, I'm contributing on a voluntary basis and because OSPOS is used by my wife's shop.
I cannot work on everything, some parts are not relevant for me like invoicing, receiving and etc. However I will contribute other functionality that would be useful.

**Comment by pablos1rvent** [comment]
@daN4cat Edit a sale would be useful xD

**Comment by bkrajendra** [comment]
@daN4cat and @RamkrishnaMondal 
I'm also very excited about this project. I'm working on bootstrap, material UI, PHP Laravel, octobercms and related domain. I'm learning the way this project is built. I'll start contributing as soon as I understand the MVC structure of OSPOS. 
Currently I'm finding it very difficult to understand.
Still somehow I manage to get value from DB and display it as purchase_tax. :)

**Comment by bkrajendra** [comment]
Now I'm Hunting for how to save it to database. :( 
I guess... I'll be spending entire night ....!

**Comment by daN4cat** [comment]
@pablos1rvent what can be edited is implemented, the was discussed and agreed not fiscally correct.

**Comment by daN4cat** [comment]
@bkrajendra you need to understand a bit code ignited. To understand how things work take one simple example like item_kits or giftcard and read module, controller and view.  Then you get an understanding about the functionality.

**Comment by bkrajendra** [comment]
Yes @daN4cat currently I'm doing the same! As for now my need is to update/add purchase capture data to existing inventory i'm trying to understand it. 
I found add_item function in recvng_lib and requisition_complete function in Recv controller doing this job of receive operation. I've done necessary changes by passing $p_tax (purchase tax) argument to add_item function.

```php

       $price=$price!=null ? $price: $item_info->cost_price; 
        $p_tax=$p_tax!=null ? $p_tax: $item_info->custom1; //my change - im using custom fld1
        $item = array(($insertkey)=>
        array(
            'item_id'=>$item_id,
            'item_location'=>$item_location,
            'stock_name'=>$this->CI->Stock_location->get_location_name($item_location),
            'line'=>$insertkey,
            'name'=>$item_info->name,
            'description'=>$description!=null ? $description: $item_info->description,
            'serialnumber'=>$serialnumber!=null ? $serialnumber: '',
            'allow_alt_description'=>$item_info->allow_alt_description,
            'is_serialized'=>$item_info->is_serialized,
            'quantity'=>$quantity,
            'discount'=>$discount,
            'in_stock'=>$this->CI->Item_quantity->get_item_quantity($item_id, $item_location)->quantity,
            'price'=>$price,
            'receiving_quantity'=>$receiving_quantity!=null ? $receiving_quantity : $item_info->receiving_quantity,
            'total'=>$this->get_item_total($quantity, $price, $discount),
            'p_tax'=>$p_tax //my change
            )
        );
```

changes in requisition_complete function in Receiving controller
```php
 function requisition_complete()
    {
    	if ($this->receiving_lib->get_stock_source() != $this->receiving_lib->get_stock_destination()) 
    	{
    		foreach($this->receiving_lib->get_cart() as $item)
    		{
    			$this->receiving_lib->delete_item($item['line']);
    			$this->receiving_lib->add_item($item['item_id'],$item['quantity'],$this->receiving_lib->get_stock_destination(),$item['p_tax']);//added p_tax here
    			$this->receiving_lib->add_item($item['item_id'],-$item['quantity'],$this->receiving_lib->get_stock_source(),$item['p_tax']); //added p_tax here
    		}
    		
			$this->complete();
    	}
    	else 
    	{
    		$data['error']=$this->lang->line('recvs_error_requisition');
    		$this->_reload($data);	
    	}
    }

```


Its getting added to current cart and even updates to receipt. not getting how to save to db.
Its ok I'm working on it and I'll find it out..!

**Comment by pablos1rvent** [comment]
@daN4cat Is not fiscally correct? Why? In which version was implemented?

**Comment by bkrajendra** [comment]
same question here. Why changing purchase cost or retail cost is not fiscally correct? In my case my both cost changes once in a month or after 2 months. Even tax changes in a year.

**Comment by pablos1rvent** [comment]
@bkrajendra Anyways if you wrong with the sale, and you need to edit or the customer want a new item, you are not going to create other receipt or invoice for the same customer in the same day, I would like to modify that sale, i hope you can understand me,

Greetings

**Comment by bkrajendra** [comment]
@pablos1rvent You can do this currently with returns.
Select Returns and enter your receipt id (e.g. POS 19) . it will show you entire sale.
now you can delete/return any of product and correct your sale.

**Comment by odiea** [comment]
Yes that's another good option.

**Comment by pablos1rvent** [comment]
@bkrajendra Shit, that fix all, you are great :D

**Comment by RamkrishnaMondal** [comment]
@bkrajendra Loook at application/model/receiving.php around line 54

```
function save($items, $supplier_id, $employee_id, $comment, $invoice_number, $payment_type, $receiving_id=false)
	{
		if(count($items)==0)
			return -1;

		$receivings_data = array(
			'supplier_id'=>$this->Supplier->exists($supplier_id) ? $supplier_id : null,
			'employee_id'=>$employee_id,
			'payment_type'=>$payment_type,
			'comment'=>$comment,
			'invoice_number'=>$invoice_number
		);

```

**Comment by bkrajendra** [comment]
@RamkrishnaMondal I've gone through this also but I guess it updates ospos_receivings and ospos_receivings_items table not the actual item cost.

I'm not getting this. If we change the cost here , it changes item unit price in this table. but this changed price does get reflected in any sales.

**Comment by bkrajendra** [comment]
I've added function in Item model
```
	public function change_p_tax($item_id, $p_tax)
	{


		$data = array('custom1' => $p_tax);

		return $this->save($data, $item_id);
	}	
```

and used this at Save method in receiving model.
I'm getting it now step by step. Thanks @RamkrishnaMondal

**Comment by tanakornp** [comment]
If you enable "Calc avg. Price (Receiving)" in store config. It will do update you cost but average it tho.

but you can change it form line 95 of  application/models/receiving.php

`			// update cost price, if changed AND is set in config as wanted
			if($cur_item_info->cost_price != $item['price']
					AND	$this->config->item('receiving_calculate_average_price') != FALSE)
			{
				$this->Item->change_cost_price($item['item_id'],
												$items_received,
												$item['price'],
												$cur_item_info->cost_price);
			}`

Please noted that the price that were sent to calculate and update is not included the discount. For me I sent the discounted price instead. 

If you want to save only the latest cost, You have to go to 
opensourcepos/application/models/Item.php about line 458

`public function change_cost_price($item_id, $items_received, $new_price, $old_price = null)
	{
		if($old_price === null)
		{
			$item_info = $this->get_info($item_id);
			$old_price = $item_info->cost_price;
		}
		$this->db->from('item_quantities');
		$this->db->select_sum('quantity');
        $this->db->where('item_id', $item_id);
		$this->db->join('stock_locations', 'stock_locations.location_id=item_quantities.location_id');
        $this->db->where('stock_locations.deleted', 0);
		$old_total_quantity = $this->db->get()->row()->quantity;
		$total_quantity = $old_total_quantity + $items_received;
		$average_price = bcdiv(bcadd(bcmul($items_received, $new_price), bcmul($old_total_quantity, $old_price)), $total_quantity);
		$data = array('cost_price' => $average_price);
		return $this->save($data, $item_id);
	}
}` 

You can change whatever cost you want to save from here.

Not sure if this is what you want to do around the updating cost and I am pretty sure there might be smarter way to get this done. 
Anyway hope it help.

**Comment by bkrajendra** [comment]
@tanakornp you are right. but instead of making changes in existing function I've added two new.

```php

	public function change_unit_price($item_id, $unit_price)
	{


		$data = array('unit_price' => $unit_price);

		return $this->save($data, $item_id);
	}
	public function change_cost_price_from_input($item_id, $new_cost_price)
	{


		$data = array('cost_price' => $new_cost_price);

		return $this->save($data, $item_id);
	}		

```
Ive put these two methods in Item.php Model.
Then calling these in save method in Receiving.php Model
```php
			// update p_tax , if changed
			if($cur_item_info->custom1 != $item['p_tax'])
			{
				$this->Item->change_p_tax($item['item_id'], $item['p_tax']);
			}

			// update Unit Price , if changed
			if($cur_item_info->unit_price != $item['unit_price'])
			{
				$this->Item->change_unit_price($item['item_id'], $item['unit_price']);
			}

			// update Unit Price , if changed
			if($cur_item_info->cost_price != $item['price'])
			{
				$this->Item->change_cost_price_from_input($item['item_id'], $item['cost_price']);
			}
```

I hope somebody just acknowledge if this is ok. @RamkrishnaMondal , @daN4cat , @jekkos

**Comment by bkrajendra** [comment]
I'm now having purchase price and purchase tax in item db.
In sales register i'm applying this tax to purchase cost to get actual purchase cost.

Now i want to do same in receive register.
![ptax](https://cloud.githubusercontent.com/assets/994083/15181612/d6b806ac-17a5-11e6-8c2c-781b005861c5.PNG)
This what I've done. Now i need purchase tax to get added to cost and reflect it in total of each item.
But im not finding it. Ive tried a lot through  - add_item and edit_item
```php
 'total'=>$this->get_item_total($quantity, $price, $discount,$p_tax), 
```
But not happening. which function is updating this total.

**Comment by RamkrishnaMondal** [comment]
I am sorry, but what you are trying to do, is not fiscally correct. You are trying to post purchase price before-tax and sales price after-tax, hence breaking the accounting convention. Post both sides either including taxes or excluding taxes. 
For example, if you put 48.63 as cost price, 5% as tax, and 57.14 (60 divided by 1.05 to exclude 5% tax from MRP) instead of 60 every thing will be ok. Please see attached.
![screenshot from 2016-05-11 19 09 24](https://cloud.githubusercontent.com/assets/14151032/15182987/bb1bd994-17ac-11e6-9572-804ec9f69930.png)

In accounting ....
Gross Profit = Net sales - Cost of goods Sold
Cost of goods Sold  = Purchase price + Input tax + All other expenses incurred to make the goods salable (eg, carriage).

Hope this helps.

**Comment by daN4cat** [comment]
You can input sale price after tax and tick the config checkbox saying sale prices include tax.

**Comment by bkrajendra** [comment]
You are absolutely right @RamkrishnaMondal , but I want actual purchase cost and purchase tax to be in database. Every Time i'll only get a invoice from my vendor with purchase cost and purchase tax. Its impossible to update these product cost by computing it with purchase tax as the list is long.. Hence what i'm doing is to enter purchase cost without tax, enter purchase tax. then compute purchase cost with purchase tax at the time of selling. so that in every sell i'm passing computed cost. This is working fine. I'm doing this in save function in sale model.
```php
//My additions to calculate purchase cost.
            $tax_fraction1 = bcdiv($cur_item_info->custom1, 100, PRECISION);
            $tax_amount1 =  bcmul($cur_item_info->cost_price, $tax_fraction1, PRECISION);
            $cost_price = bcadd($cur_item_info->cost_price, $tax_amount1, PRECISION);

			$sales_items_data = array(
				'sale_id'			=> $sale_id,
				'item_id'			=> $item['item_id'],
				'line'				=> $item['line'],
				'description'		=> character_limiter($item['description'], 30),
				'serialnumber'		=> character_limiter($item['serialnumber'], 30),
				'quantity_purchased'=> $item['quantity'],
				'discount_percent'	=> $item['discount'],
				'item_cost_price'	=> $cost_price,//$cur_item_info->cost_price,
				'item_unit_price'	=> $item['price'],
				'item_location'		=> $item['item_location']
			);
```

Samething I want to do while doing receive. I just want to compute a purchase invoice from purchase tax and purchase cost. ... but im not able to find code block which updates total cost for each item.

**Comment by daN4cat** [comment]
That will be fine.

By the way, what are we trying to achieve here? Have correct profit?

From tax payment point of view you pay the tax of what sold taken the tax of what you were invoiced by suppliers in the given time (quarter in the UK). So this refined figure is just to get a correct feeling of the profit and loss, not to get a cash flow view.

**Comment by bkrajendra** [comment]
In addition to profit what i care more is usability - user interface for practical use. 
Anyways .. Ive got the way to update total. It was in "complete" method. Actual calculate method is in receiving_lib - get_total
my modifications:
```php
  function get_total()
    {
        $total = 0;
        foreach($this->get_cart() as $item)
        {
            $total = bcadd($total, $this->get_item_total($item['quantity'], $item['price'], $item['discount'],$item['p_tax']), PRECISION);
        }
        
        return $total;
    }
```
added p_tax and its fixed.
Now only thing left is its not reflecting in reports. I guess it due to temporary table.
Any thoughts.

**Comment by daN4cat** [comment]
Yes, you need to add it to the temporary table.

I'm not getting the usability point, it's not an user interface issue, it's an accountancy requirement so a functionality matter. Usability is something else.

**Comment by daN4cat** [comment]
@bkrajendra is this something you will contribute back when finished?

**Comment by bkrajendra** [comment]
yes @daN4cat  i'm definitely doing it to contribute. I will be really happy if my little contributions can help build a good software.

**Comment by daN4cat** [comment]
Good and looking forward to seeing this new piece of work.

By the way be aware that I need to do some changes for the XSS clean and fix the "invoice" field to be "Reference" in receiving in the next days.
So you will need to merge in any change.

**Comment by daN4cat** [comment]
@bkrajendra hope you could keep up with all the changes, but now we are maturing 3.0 and we should be close to get to an end and release it.

**Comment by daN4cat** [comment]
@bkrajendra have you ever progressed this? Willing to contribute?

**Comment by sanubari97** [comment]
and please add Value Added Tax (VAT).

**Comment by bkrajendra** [comment]
Sorry @daN4cat , but I never got time latter to work on this. I dont think its a standard way.
It really need a dedicated time to seriously complete this functionality of receiving/purchase module.
This is a very important part of any POS. I'm preoccupied with my embedded system projects, so it's difficult for me to work on it.

**Comment by SteveIreland** [comment]
Item cost update is going to be part of the purchasing development so please add what should happen regarding costs ... particularly any special VAT rules ... to the wiki

**Comment by objecttothis** [comment]
#1563 was marked as a duplicate, but is slightly different.  Nonetheless I'll add it here.  Cost Price in receivings currently does not change the cost price in the Items table.  Cost Price entered into Receivings of existing items should do one of two different things
1- It should average the existing cost price with the receivings cost price and update the database with the new average.  For example, if There are 10 widgets in stock at a wholesale (cost price) of $1 and I do receivings for 15 of that same item at a cost price of $0.50 should change the cost price in the items table to $0.70 (15*0.50) + (10 * 1) / 25 which is the average of the cost prices (This is preferable and I think the correct way but open to debate).
2- It could simply replace the cost price in the items table with the new cost price.  This isn't preferable because it doesn't take into account the cost of items already in stock.

**Comment by objecttothis** [comment]
@SteveIreland I assigned you because I think you are working on purchasing.  3.4.0 is not written in stone, so whatever release Purchasing is slated for can be the same for this.

**Comment by SteveIreland** [comment]
Makes sense to me.

**Comment by izeblood** [comment]
> #1563 was marked as a duplicate, but is slightly different. Nonetheless I'll add it here. Cost Price in receivings currently does not change the cost price in the Items table. Cost Price entered into Receivings of existing items should do one of two different things
> 1- It should average the existing cost price with the receivings cost price and update the database with the new average. For example, if There are 10 widgets in stock at a wholesale (cost price) of $1 and I do receivings for 15 of that same item at a cost price of $0.50 should change the cost price in the items table to $0.70 (15*0.50) + (10 * 1) / 25 which is the average of the cost prices (This is preferable and I think the correct way but open to debate).
> 2- It could simply replace the cost price in the items table with the new cost price. This isn't preferable because it doesn't take into account the cost of items already in stock.

or add another option instead of Calc avg. Price (Receiving) how about the highest price may always retain as cost price?
